pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,268 @@
+			let instance = Self { id };
+			let contract_id = instance.env().account_id();
+			api::create(id, contract_id, min_balance).map_err(Psp22Error::from)?;
+			instance
+				.env()
+				.emit_event(Created { id, creator: contract_id, admin: contract_id });
+			Ok(instance)
 		}
+	}
 
+	impl Psp22 for Fungible {
+		/// Returns the total token supply.
 		#[ink(message)]
-		pub fn balance_of(&self, token: TokenId, owner: AccountId) -> Result<Balance> {
-			api::balance_of(token, owner)
+		fn total_supply(&self) -> Balance {
+			api::total_supply(self.id).unwrap_or_default()
 		}
 
+		/// Returns the account balance for the specified `owner`.
+		///
+		/// # Parameters
+		/// - `owner` - The account whose balance is being queried.
 		#[ink(message)]
-		pub fn allowance(
-			&self,
-			token: TokenId,
-			owner: AccountId,
-			spender: AccountId,
-		) -> Result<Balance> {
-			api::allowance(token, owner, spender)
+		fn balance_of(&self, owner: AccountId) -> Balance {
+			api::balance_of(self.id, owner).unwrap_or_default()
 		}
 
+		/// Returns the allowance for a `spender` approved by an `owner`.
+		///
+		/// # Parameters
+		/// - `owner` - The account that owns the tokens.
+		/// - `spender` - The account that is allowed to spend the tokens.
 		#[ink(message)]
-		pub fn transfer(&mut self, token: TokenId, to: AccountId, value: Balance) -> Result<()> {
-			api::transfer(token, to, value)
+		fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+			api::allowance(self.id, owner, spender).unwrap_or_default()
 		}
 
+		/// Transfers `value` amount of tokens from the caller's account to account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data in unspecified format.
 		#[ink(message)]
-		pub fn transfer_from(
+		fn transfer(
 			&mut self,
-			token: TokenId,
-			from: AccountId,
 			to: AccountId,
 			value: Balance,
 			_data: Vec<u8>,
-		) -> Result<()> {
-			api::transfer_from(token, from, to, value)
+		) -> Result<(), Psp22Error> {
+			let caller = self.env().caller();
+			// No-op if the caller and `to` is the same address or `value` is zero.
+			if caller == to || value == 0 {
+				return Ok(());
+			}
+			api::transfer(self.id, to, value).map_err(Psp22Error::from)?;
+			self.env().emit_event(Transfer { from: Some(caller), to: Some(to), value });
+			Ok(())
 		}
 
+		/// Transfers `value` tokens on behalf of `from` to the account `to`
+		/// with additional `data` in unspecified format.
+		///
+		/// # Parameters
+		/// - `from` - The account from which the token balance will be withdrawn.
+		/// - `to` - The recipient account.
+		/// - `value` - The number of tokens to transfer.
+		/// - `data` - Additional data with unspecified format.
 		#[ink(message)]
-		pub fn approve(
+		fn transfer_from(
```

Comment:
I think this needs a note in the function doc comments to say that from needs to have pre-approved the contract (the origin of the api call) to be able to transfer the specified amount.